File: zig/src/cpu.zig
/// CPU simulator
const std = @import("std");
const constants = @import("constants.zig");
const CpuError = @import("error.zig").CpuError;
const Memory = @import("memory.zig").Memory;
const Opcode = @import("opcode.zig").Opcode;

pub const Cpu = struct {
    memory: Memory,
    timer_interval: usize,

    // Registers
    ir: ?Opcode,
    pc: usize,
    sp: usize,
    ac: i32,
    x: i32,
    y: i32,

    // State
    mode: constants.Mode,
    cycles: usize,
    interrupts_enabled: bool,

    // Output
    stdout: std.fs.File,

    // Random Number Generation
    prng: std.Random.DefaultPrng,

    pub fn init(memory: Memory, timer_interval: usize) Cpu {
        return .{
            .memory = memory,
            .timer_interval = timer_interval,
            .ir = null,
            .pc = @intFromEnum(constants.ProgramLoad.user),
            .sp = @intFromEnum(constants.StackBase.user),
            .ac = 0,
            .x = 0,
            .y = 0,
            .mode = .user,
            .cycles = 0,
            .interrupts_enabled = true,
            .stdout = std.fs.File{ .handle = std.posix.STDOUT_FILENO },
            .prng = std.Random.DefaultPrng.init(0),
        };
    }

    /// Run the program to completion
    pub fn run(self: *Cpu) !void {
        while (true) {
            const continuing = try self.step();
            if (!continuing) return;
        }
    }

    /// Execute one instruction. Returns true to continue, false on END.
    pub fn step(self: *Cpu) !bool {
        if (self.fireTimer()) {
            try self.doInterrupt(.timer);
        }

        const raw = try self.load(self.pc);
        const opcode = try Opcode.fromInt(raw);
        self.ir = opcode;

        const operand: ?i32 = if (opcode.hasOperand())
            try self.load(self.pc + 1)
        else
            null;

        const continuing = try self.execute(opcode, operand);

        self.cycles += 1;

        if (continuing and !opcode.isCti()) {
            self.pc += if (operand != null) 2 else 1;
        }

        return continuing;
    }

    fn fireTimer(self: *const Cpu) bool {
        return self.cycles > 0 and
            self.timer_interval > 0 and
            self.cycles % self.timer_interval == 0;
    }

    fn load(self: *const Cpu, address: usize) CpuError!i32 {
        if (self.mode == .user and address > @intFromEnum(constants.StackBase.user)) {
            return CpuError.SegmentationFault;
        }
        return self.memory.read(address);
    }

    fn store(self: *Cpu, address: usize, value: i32) CpuError!void {
        if (self.mode == .user and address > @intFromEnum(constants.StackBase.user)) {
            return CpuError.SegmentationFault;
        }
        return self.memory.write(address, value);
    }

    fn push(self: *Cpu, value: i32) CpuError!void {
        self.sp -= 1;
        try self.store(self.sp, value);
    }

    fn pop(self: *Cpu) CpuError!i32 {
        const stackbase = @intFromEnum(constants.StackBase.forMode(self.mode));
        if (self.sp >= stackbase) {
            return CpuError.StackUnderflow;
        }
        const value = try self.load(self.sp);
        self.sp += 1;
        return value;
    }

    /// Write an integer to stdout
    fn putInt(self: *Cpu, value: i32) void {
        var buf: [32]u8 = undefined;
        const s = std.fmt.bufPrint(&buf, "{d}", .{value}) catch return;
        self.stdout.writeAll(s) catch {};
    }

    /// Write a single character to stdout
    fn putChar(self: *Cpu, value: i32) void {
        const byte = [1]u8{@intCast(@as(u32, @bitCast(value)) & 0xFF)};
        self.stdout.writeAll(&byte) catch {};
    }

    /// Execute a decoded instruction. Returns false on END.
    ///
    /// TODO: This is yours to fill in, Erik. The switch skeleton is here
    /// with a few examples to show the pattern. The rest are marked with
    /// @panic("TODO") so the compiler will tell you which ones are left.
    ///
    /// Hint: the Rust version in ../rust/src/cpu.rs has the complete logic
    /// if you get stuck on any particular opcode's semantics.
    fn execute(self: *Cpu, opcode: Opcode, operand: ?i32) !bool {
        switch (opcode) {
            // === LOADS (examples filled in) ===
            .loadv => {
                self.ac = operand.?;
            },
            .loada => {
                self.ac = try self.load(@intCast(operand.?));
            },
            .loadi => {
                const addr = try self.load(@intCast(operand.?));
                self.ac = try self.load(@intCast(addr));
            },
            .loadx => {
                //@panic("TODO: load value at address + X into AC"),
                const addr = self.x + operand.?;
                self.ac = try self.load(@intCast(addr));
            },
            .loady => {
                //@panic("TODO: load value at address + Y into AC"),
                const addr = self.y + operand.?;
                self.ac = try self.load(@intCast(addr));
            },
            .loadspx => {
                //@panic("TODO: load value at SP + X into AC"),
                const addr = self.sp +% @as(usize, @intCast(self.x));
                self.ac = try self.load(@intCast(addr));
            },

            // === STORE ===
            .store => {
                // @panic("TODO: store AC to address"),
                const addr = @as(usize, @intCast(operand.?));
                try self.store(addr, self.ac);
            },

            // === I/O ===
            .get => {
                //@panic("TODO: random int 1-100 into AC"),
                self.ac = self.prng.random().intRangeAtMost(i32, 1, 100);
            },
            .put => {
                // port 1 = integer, port 2 = character
                const port = operand.?;
                switch (port) {
                    1 => self.putInt(self.ac),
                    2 => self.putChar(self.ac),
                    else => return CpuError.InvalidOperand,
                }
            },

            // === ARITHMETIC ===
            .addx => {
                //@panic("TODO: add X to AC"),
                self.ac += self.x;
            },
            .addy => {
                //@panic("TODO: add Y to AC"),
                self.ac += self.y;
            },
            .subx => {
                //@panic("TODO: subtract X from AC"),
                self.ac -= self.x;
            },
            .suby => {
                //@panic("TODO: subtract Y from AC"),
                self.ac -= self.y;
            },
            // === REGISTER COPIES ===
            .copytox => {
                //@panic("TODO: copy AC to X"),
                self.x = self.ac;
            },
            .copyfromx => {
                //@panic("TODO: copy X to AC"),
                self.ac = self.x;
            },
            .copytoy => {
                //@panic("TODO: copy AC to Y"),
                self.y = self.ac;
            },
            .copyfromy => {
                //@panic("TODO: copy Y to AC"),
                self.ac = self.y;
            },
            .copytosp => {
                //@panic("TODO: copy AC to SP"),
                self.sp = @intCast(self.ac);
            },
            .copyfromsp => {
                //@panic("TODO: copy SP to AC"),
                self.ac = @intCast(self.sp);
            },
            // === CONTROL FLOW ===
            .jump => {
                self.pc = @intCast(operand.?);
            },
            .jumpeq => {
                //@panic("TODO: jump if AC == 0, else pc += 2"),
                if (self.ac == 0) {
                    self.pc = @intCast(operand.?);
                } else {
                    self.pc += 2;
                }
            },
            .jumpne => {
                //@panic("TODO: jump if AC != 0, else pc += 2"),
                if (self.ac != 0) {
                    self.pc = @intCast(operand.?);
                } else {
                    self.pc += 2;
                }
            },
            .call => {
                //@panic("TODO: push return addr (pc+2), jump to operand"),
                try self.push(@intCast(self.pc));
                self.pc = @intCast(operand.?);
            },
            .ret => {
                //@panic("TODO: pop return addr, jump to it"),
                const addr = try self.pop();
                self.pc = @intCast(addr);
            },

            // === STACK ===
            .incx => {
                //@panic("TODO: increment X"),
                self.x += 1;
            },
            .decx => {
                //@panic("TODO: decrement X"),
                self.x -= 1;
            },
            .push => {
                //@panic("TODO: push AC onto stack"),
                try self.push(@intCast(self.ac));
            },
            .pop => {
                //@panic("TODO: pop stack into AC"),
                self.ac = try self.pop();
            },

            // === INTERRUPTS ===
            .interrupt => {
                try self.doInterrupt(.interrupt);
            },
            .ireturn => {
                try self.doIreturn();
            },

            // === END ===
            .end => return false,

            // Non-exhaustive enum catch-all
            _ => return CpuError.InvalidOpcode,
        }
        return true;
    }

    fn doInterrupt(self: *Cpu, target: constants.ProgramLoad) CpuError!void {
        if (!self.interrupts_enabled) {
            self.pc += 1;
            return;
        }

        self.mode = .system;
        self.interrupts_enabled = false;

        const u_sp = self.sp;
        var u_pc = self.pc;

        if (target == .interrupt) {
            u_pc += 1;
        }

        self.sp = @intFromEnum(constants.StackBase.system);
        try self.push(@intCast(u_sp));
        try self.push(@intCast(u_pc));

        self.pc = @intFromEnum(target);
    }

    fn doIreturn(self: *Cpu) CpuError!void {
        self.pc = @intCast(try self.pop());
        self.sp = @intCast(try self.pop());
        self.interrupts_enabled = true;
        self.mode = .user;
    }
};

File: rust/src/cpu.rs
/// CPU simulator
use rand::Rng;

use crate::constants::{Mode, ProgramLoad, StackBase};
use crate::error::{CpuError, Result};
use crate::memory::Memory;
use crate::opcode;

pub struct Cpu {
    pub memory: Memory,
    pub timer_interval: usize,

    // Registers
    pub ir: i32,
    pub pc: usize,
    pub sp: usize,
    pub ac: i32,
    pub x: i32,
    pub y: i32,

    // State
    pub mode: Mode,
    pub cycles: usize,
    pub interrupts_enabled: bool,
}

impl Cpu {
    pub fn new(memory: Memory, timer_interval: usize) -> Self {
        Self {
            memory,
            timer_interval,
            ir: 0,
            pc: ProgramLoad::User as usize,
            sp: StackBase::User as usize,
            ac: 0,
            x: 0,
            y: 0,
            mode: Mode::User,
            cycles: 0,
            interrupts_enabled: true,
        }
    }

    pub fn reset(&mut self) {
        self.mode = Mode::User;
        self.ir = 0;
        self.pc = ProgramLoad::User as usize;
        self.sp = StackBase::User as usize;
        self.ac = 0;
        self.x = 0;
        self.y = 0;
        self.cycles = 0;
        self.interrupts_enabled = true;
    }

    /// Run the program to completion
    pub fn run(&mut self) -> Result<()> {
        loop {
            match self.step() {
                Ok(false) => return Ok(()),  // END instruction
                Ok(true) => continue,
                Err(e) => return Err(e),
            }
        }
    }

    /// Execute one instruction. Returns Ok(true) to continue, Ok(false) on END.
    pub fn step(&mut self) -> Result<bool> {
        // Check timer interrupt
        if self.fire_timer() {
            self.interrupt(ProgramLoad::Timer)?;
        }

        // Fetch instruction
        self.ir = self.load(self.pc)?;
        let info = opcode::lookup(self.ir)?;

        // Fetch operand if needed
        let operand = if info.has_operand {
            Some(self.load(self.pc + 1)?)
        } else {
            None
        };

        // Execute
        let is_cti = info.is_cti;
        let advanced = self.execute(info.name, operand)?;

        self.cycles += 1;

        // Advance PC if the instruction didn't do it (not a CTI)
        if !is_cti && advanced {
            self.pc += if operand.is_some() { 2 } else { 1 };
        }

        Ok(advanced)
    }

    fn fire_timer(&self) -> bool {
        self.cycles > 0
            && self.timer_interval > 0
            && self.cycles % self.timer_interval == 0
    }

    fn load(&self, address: usize) -> Result<i32> {
        if self.mode == Mode::User {
            let user_end = StackBase::User as usize;
            if address > user_end {
                return Err(CpuError::SegmentationFault(format!("load from {address}")));
            }
        }
        self.memory.read(address)
    }

    fn store(&mut self, address: usize, value: i32) -> Result<()> {
        if self.mode == Mode::User {
            let user_end = StackBase::User as usize;
            if address > user_end {
                return Err(CpuError::SegmentationFault(format!("store to {address}")));
            }
        }
        self.memory.write(address, value)
    }

    fn push(&mut self, value: i32) -> Result<()> {
        self.sp -= 1;
        self.store(self.sp, value)
    }

    fn pop(&mut self) -> Result<i32> {
        let stackbase = StackBase::for_mode(self.mode) as usize;
        if self.sp >= stackbase {
            return Err(CpuError::StackUnderflow(self.sp, self.mode));
        }
        let value = self.load(self.sp)?;
        self.sp += 1;
        Ok(value)
    }

    /// Execute a single instruction by name. Returns false on END.
    fn execute(&mut self, name: &str, operand: Option<i32>) -> Result<bool> {
        match name {
            "loadv"      => { self.ac = operand.unwrap(); }
            "loada"      => { self.ac = self.load(operand.unwrap() as usize)?; }
            "loadi"      => {
                let addr = self.load(operand.unwrap() as usize)?;
                self.ac = self.load(addr as usize)?;
            }
            "loadx"      => { self.ac = self.load((operand.unwrap() + self.x) as usize)?; }
            "loady"      => { self.ac = self.load((operand.unwrap() + self.y) as usize)?; }
            "loadspx"    => { self.ac = self.load(self.sp + self.x as usize)?; }
            "store"      => { self.store(operand.unwrap() as usize, self.ac)?; }
            "get"        => { self.ac = rand::rng().random_range(1..=100); }
            "put"        => {
                match operand.unwrap() {
                    1 => print!("{}", self.ac),
                    2 => print!("{}", char::from(self.ac as u8)),
                    p => return Err(CpuError::InvalidOperand(format!("unknown port={p}"))),
                }
            }
            "addx"       => { self.ac += self.x; }
            "addy"       => { self.ac += self.y; }
            "subx"       => { self.ac -= self.x; }
            "suby"       => { self.ac -= self.y; }
            "copytox"    => { self.x = self.ac; }
            "copyfromx"  => { self.ac = self.x; }
            "copytoy"    => { self.y = self.ac; }
            "copyfromy"  => { self.ac = self.y; }
            "copytosp"   => { self.sp = self.ac as usize; }
            "copyfromsp" => { self.ac = self.sp as i32; }
            "jump"       => { self.pc = operand.unwrap() as usize; }
            "jumpeq"     => {
                if self.ac == 0 {
                    self.pc = operand.unwrap() as usize;
                } else {
                    self.pc += 2;
                }
            }
            "jumpne"     => {
                if self.ac != 0 {
                    self.pc = operand.unwrap() as usize;
                } else {
                    self.pc += 2;
                }
            }
            "call"       => {
                self.push(self.pc as i32 + 2)?;
                self.pc = operand.unwrap() as usize;
            }
            "ret"        => {
                self.pc = self.pop()? as usize;
            }
            "incx"       => { self.x += 1; }
            "decx"       => { self.x -= 1; }
            "push"       => { self.push(self.ac)?; }
            "pop"        => { self.ac = self.pop()?; }
            "interrupt"  => { self.interrupt(ProgramLoad::Interrupt)?; }
            "ireturn"    => { self.ireturn()?; }
            "end"        => { return Ok(false); }
            _            => { return Err(CpuError::MachineCheck(format!("Unknown instruction: {name}"))); }
        }
        Ok(true)
    }

    fn interrupt(&mut self, program_load: ProgramLoad) -> Result<()> {
        if !self.interrupts_enabled {
            self.pc += 1;
            return Ok(());
        }

        self.mode = Mode::System;
        self.interrupts_enabled = false;

        let u_sp = self.sp;
        let mut u_pc = self.pc;

        if program_load == ProgramLoad::Interrupt {
            u_pc += 1;
        }

        self.sp = StackBase::System as usize;
        self.push(u_sp as i32)?;
        self.push(u_pc as i32)?;

        self.pc = program_load as usize;
        Ok(())
    }

    fn ireturn(&mut self) -> Result<()> {
        self.pc = self.pop()? as usize;
        self.sp = self.pop()? as usize;
        self.interrupts_enabled = true;
        self.mode = Mode::User;
        Ok(())
    }
}
